### PR TITLE
[opt](brpc) check and remove unavailable brpc stubs

### PR DESCRIPTION
**Pull request by mrhhsg**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Check List (For Committer)

- Test 

    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No colde files have been changed.
        - [ ] Other reason 

- Behavior changed:

    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?

    - [ ] No.
    - [ ] Yes. 

- Release note

    
    
    None

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

File: be/src/runtime/fragment_mgr.cpp [hunk 1]
```diff
@@ -909,6 +911,7 @@ void FragmentMgr::cancel_worker() {
             c->clear_finished_tasks();
         }
 
+        std::unordered_map<TNetworkAddress, BrpcItem> brpc_stub_with_queries;
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp:914]: 不要这么写，会出现2个rpc 链接的host 一样，但是brpc的对象不一样的问题。
可以使用brpc stub的指针做key，然后把network address 放到BrpcItem 这里面，供check的时候打log 使用。

File: be/src/runtime/fragment_mgr.cpp [hunk 2]
```diff
@@ -1034,10 +1050,55 @@ void FragmentMgr::cancel_worker() {
...
+        request.set_hello(message);
+        PHandShakeResponse response;
+        brpc::Controller cntl;
+        cntl.set_timeout_ms(500);
+        brpc_item.brpc_stub->hand_shake(&cntl, &request, &response, nullptr);
+
+        if (cntl.Failed()) {
+            error_message = cntl.ErrorText();
+            LOG(WARNING) << "brpc stub: " << network_address << " check failed: " << error_message;
+        } else if (response.has_status() && response.has_hello() && response.hello() == message &&
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp:1074]: 直接检查是否ok 就行了，不用比对消息体对不对

File: be/src/runtime/fragment_mgr.cpp [hunk 3]
```diff
@@ -1034,10 +1050,55 @@ void FragmentMgr::cancel_worker() {
...
+                                        BrpcItem& brpc_item) {
+    const std::string message = "hello doris!";
+    std::string error_message;
+    int32_t failed_count = 0;
+    while (true) {
+        PHandShakeRequest request;
+        request.set_hello(message);
+        PHandShakeResponse response;
+        brpc::Controller cntl;
+        cntl.set_timeout_ms(500);
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp:1068]: 500 * (failed_count + 1)

File: be/src/runtime/fragment_mgr.cpp [hunk 4]
```diff
@@ -1034,10 +1050,55 @@ void FragmentMgr::cancel_worker() {
...
 
+void FragmentMgr::_check_brpc_available(const TNetworkAddress& network_address,
+                                        BrpcItem& brpc_item) {
+    const std::string message = "hello doris!";
+    std::string error_message;
+    int32_t failed_count = 0;
+    while (true) {
+        PHandShakeRequest request;
+        request.set_hello(message);
+        PHandShakeResponse response;
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp:1066]: should also set max try = 10

File: be/src/runtime/fragment_mgr.h [hunk 5]
```diff
@@ -148,6 +148,11 @@ class FragmentMgr : public RestMonitorIface {
     std::shared_ptr<QueryContext> get_or_erase_query_ctx_with_lock(const TUniqueId& query_id);
 
 private:
+    struct BrpcItem {
+        std::shared_ptr<PBackendService_Stub> brpc_stub;
+        std::map<std::string, std::weak_ptr<QueryContext>> queries;
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.h:153]: why it is a map? a vector is not ok?

**Review by yiguolei — DISMISSED** [review]

**Comment by yiguolei** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

File: be/src/runtime/fragment_mgr.cpp [hunk 6]
```diff
@@ -912,9 +915,18 @@ void FragmentMgr::cancel_worker() {
                         LOG_WARNING("Query {} is timeout", print_id(it->first));
                         queries_timeout.push_back(it->first);
                         ++it;
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp:917]: 这个it ++ 是多余的

**Review by yiguolei — APPROVED** [review]

**Review by yiguolei — APPROVED** [review]

**Comment by yiguolei** [comment]
run buildall

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Pull request merged**